Migrate GalleryFragment and GardenFragment to Compose

arriolac opened the pull request:
Resolves #816 #814
Fixes #773

arriolac commented:
@mlykotom am I using the `ReportDrawnWhen` correctly in [3903c31](https://github.com/android/sunflower/pull/819/commits/3903c318f5b7c7b3beddd2bc194c4afb8b41452f)?

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/garden/GardenScreen.kt:
```diff
@@ -0,0 +1,263 @@
+    gardenPlants: List<PlantAndGardenPlantings>,
+    onPlantClick: (PlantAndGardenPlantings) -> Unit
+) {
+    LazyVerticalGrid(
```
alternative is to call ReportDrawnWhen the grid has rendered some items
``` kotlin
val gridState = rememberLazyGridState()
ReportDrawnWhen { gridState.layoutInfo.totalItemsCount > 0 }
```

This way you'd report it more closer to when the UI is actually ready

arriolac replied:
I like that. Going to move the `ReportDrawnWhen` calls withing `EmptyGarden` and `GardenList`

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/garden/GardenScreen.kt:
```diff
@@ -0,0 +1,263 @@
+                text = vm.plantName,
+                Modifier
+                    .padding(
+                        top = dimensionResource(id = R.dimen.margin_normal),
```
would it be worth to extract the dimension into a variable to prevent parsing it multiple times?

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/plantdetail/PlantDetailView.kt:
```diff
@@ -315,7 +330,7 @@ private fun PlantImage(
                     e: GlideException?,
                     model: Any?,
-                    target: Target<Drawable>?,
+                    target: com.bumptech.glide.request.target.Target<Drawable>?,
```
can we import the Target instead of the full qualifier here? (same 👇🏼 )

mlykotom commented on app/src/main/res/values/styles.xml:
```diff
@@ -20,7 +20,8 @@
         <item name="colorPrimary">@color/sunflower_green_500</item>
         <item name="colorPrimaryVariant">@color/sunflower_green_700</item>
-        <item name="colorOnPrimary">@color/sunflower_yellow_500</item>
+<!--        <item name="colorOnPrimary">@color/sunflower_yellow_500</item>-->
```
remove?

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/plantdetail/PlantDetailView.kt:
```diff
@@ -551,12 +584,14 @@ private fun PlantDescription(description: String) {
 @Preview
 @Composable
 private fun PlantDetailContentPreview() {
+    // FIXME: Preview is broken because of GlideImage. See: https://github.com/bumptech/glide/issues/4977
```
one thing we could do is wrap GlideImage in our own composable and use `LocalInspectionMode` to render e.g. red Box .. this way, the previews are working until the issue is resolved

arriolac replied:
Good idea! Will create a wrapper for `GlideImage`

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/garden/GardenScreen.kt:
```diff
@@ -0,0 +1,263 @@
+    PreviewParameterProvider<List<PlantAndGardenPlantings>> {
+    override val values: Sequence<List<PlantAndGardenPlantings>> =
+        sequenceOf(
+            emptyList(),
```
is it needed?

arriolac replied:
Yep! This way we can preview both populated and empty states

mlykotom replied:
I see, I didn't realise that, thanks for explaining!

mlykotom reviewed (commented):
I have some comments/questions. ReportDrawn could be more readable. Thanks!

arriolac commented:
> I have some comments/questions. ReportDrawn could be more readable. Thanks!

Thanks for your review, @mlykotom ! I addressed your feedback in [c6ae463](https://github.com/android/sunflower/pull/819/commits/c6ae4630b85db927c71c8bf61708682d4db9e50c)

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/garden/GardenScreen.kt:
```diff
@@ -0,0 +1,265 @@
+
+@Composable
+private fun GardenScreen(
+    gardenPlants: List<PlantAndGardenPlantings>?,
```
this list won't be null anymore, right? because you pass `initial = emptyList()`

mlykotom commented on app/src/main/java/com/google/samples/apps/sunflower/compose/garden/GardenScreen.kt:
```diff
@@ -0,0 +1,265 @@
+
+@Composable
+private fun EmptyGarden(onAddPlantClick: () -> Unit) {
+    Column(
```
You missed the report drawn here.
Depending on when you check the issue, today is scheduled the alpha03 which contains `ReportDrawn()`, so you may use it instead of the `ReportDrawnWhen { true }`

```suggestion
    ReportDrawnWhen { true}
    Column(
```

arriolac replied:
Thanks - I wasn't sure if `ReportDrawnWhen { true }` is needed in this case since it's always `true`.

mlykotom reviewed (approved):
Some nits, otherwise lgtm
